### ISSUE TITLE
Latent averaging to the logits?

### ISSUE THREAD

**Issue by neonbjb** [lucidrains/perceiver-pytorch#11]
I read through the paper last night and came away confused about a few things. I looked through your code hoping for some clarity.

One issue that doesn't seem to be explained in the paper (or I am missing it) is how the authors go from a set of latents to the logits used at the classification head. You implemented this by taking the mean of the latent set:

https://github.com/lucidrains/perceiver-pytorch/blob/main/perceiver_pytorch/perceiver_pytorch.py#L203

Is this actually how the authors convert to logits?

**Comment by lucidrains** [comment]
@neonbjb I have no clue, but coming from NLP, meaning the last layer of embeddings or using a CLS token rarely makes a difference

**Comment by neonbjb** [comment]
Alright that's fine, I just wanted to make sure I wasn't missing something. I am betting you are right that either this (averaging) or using a CLS token is what the original authors did.

The thing is - this doesn't seem like a very good training strategy. In BERT-style models, the majority of the training takes place in a regime where all of the output latents matter equally but they are also distinct from each other. Only in the last step do you tell the model to throw all that distinction out and converge to a single classification task. I would be very surprised if training a model from scratch with converged latents like this produces the optimal results. As stated, I don't think this is a problem with your implementation.

**Comment by lucidrains** [comment]
@neonbjb I don't think it matters. An extra layer or two of attention and the problem is solved https://arxiv.org/abs/2102.11972

**Comment by lucidrains** [comment]
@neonbjb traditionally the CLS token is there so that it is easy to conduct masked token training

**Comment by lucidrains** [comment]
https://twitter.com/ak92501/status/1373797875390545924?s=20

**Comment by neonbjb** [comment]
Just to close this out - I contacted the authors and they confirmed that your implementation is correct - they do, in fact, take the mean across the latent array.

Thanks for all your hard work with these repos!

**Comment by lucidrains** [comment]
@neonbjb awesome! thanks for looking into it :)
